## Lab 4: Coding region and other mRNA features

##### Q0: What's your name?

The purpose of this exercise is to retrieve your gene's mRNA sequence from Genbank and perform some preliminary analysis. You'll parse your gene using the "Features" annotations of its Genbank record.

##### Q1: What gene have you chosen for your final project?

### Identifying coding and untranlated regions of your mRNA.

In [8]:
from Bio import Entrez # we'll import a set of code written for common biological applications
from Bio import SeqIO

In [9]:
GI_id = "" # enter the accession # or GeneID for the mRNA sequence of your gene ***as a string***

Documentation for efecth http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EFetch  

Code adapted from Biopython Tutorial, 5.3.1  Parsing GenBank records from the net

In [1]:
Entrez.email = "" # let NCBI know who you are
handle = Entrez.efetch(db="nucleotide", rettype="gb", retmode="text", id=GI_id) # db = database to query, rettype = retrieval type, retmode = retrieval mode 
seq_record = SeqIO.read(handle, "gb") #read the contents "fetched" from NCBI
handle.close()

print #print the contents of "seq_record"
print #print the type of data in "seq_record"

#### Q2: How many "features" does your gene have?

Open up a web browser and navigate to NCBI Nucleotide. Perform a search for your gene. When you click to see the record for your gene a Genbank record is displayed. Scroll down until you see the "Features" section. 

**Q3: What type of information is found here?**

Locate the feature called "CDS". 

**Q4: What does "CDS" mean?**

Next we'll loop over the features of your gene's genbank record and print each feature.

In [2]:
for f in seq_record.features: # for each feature "f" in this record's features
    # display the feature

**Q5: What "type" of features are returned? You know know about the "CDS" feature - are there other features you're unfamiliar with? If so look up those feature types here: http://www.insdc.org/documents/feature_table.html and make note below of the type of information each contains.**

Notice how each of the features has some shared descriptors: type, location, and qualifiers. For a given feature you can access the informtion in these descriptors directly.   
For example, if I wanted to loop over all the features in a record and print out the type of each feature I would use a loop like this:

In [12]:
for f in seq_record.features: # for each feature in this records features
    print f.type # print out the type of feature

source
gene
misc_feature
CDS
sig_peptide
misc_feature
mat_peptide
misc_feature
STS


If I wanted to print out the location of each feature I might use this loop:

In [13]:
for f in seq_record.features: # for each feature in this records features
    print f.location # print out the type of feature

[0:4378](+)
[0:4378](+)
[288:291](+)
[297:1092](+)
[297:369](+)
[369:1020](+)
[369:1020](+)
[456:606](+)
[401:813](+)


Using the above two loops as a guide - write a for loop to print the location if the feature type is CDS. Since we're using an mRNA sequence, there should be only one CDS region in this Genbank record. **If there is more than one CDS feature listed in the Genbank record, see your TA before continuing on with the exercise.**

In [3]:
for f in seq_record.features: # for each feature in this records features
    # ask if the feature is a CDS
    # if so print out the location of the feature

**Q6: Compare the location of the CDS returned by Biopython to the location of the CDS according to the NCBI Nucleotide record for your gene. Are they the same? Explain any differences you see.**  

Hint: Read the Biopython cookbook (http://biopython.org/DIST/docs/tutorial/Tutorial.html) section 4.3.3,  Sequence described by a feature or location.

#### **Q7: Does the CDS region start at the first nucleotide of your mRNA sequence? Is the last nucleotide of the CDS region the last nucleotide of your mRNA sequence?   What is the region of the mRNA before the CDS called? What is the region of the mRNA following the CDS called?**

The location descriptor has some additional characteristics which are very useful. Suppose we want to retrieve the nucleotide sequence of CDS and assign it to a variable called CDS_seq. Notice that the CDS feature does not actually contain the nucleotide sequence of the CDS. Since the location of the CDS is defined relative to the sequence of the original record we can use the information in "location" to return the slice of the original sequence corresponding to the CDS. Since we've used an mRNA query, we don't have to deal with piecing together exons - this has already been done for us. 

But how do we retreive the begining and ending position from \[###:###](+)? We'll see in later labs that we can use regular expressions to pieces of informaiton out of a string without knowing beforehand the position of the information within the string. But we don't have to do that here. 

For example, if f.location is \[297:369](+)  
f.location.start will return 297.  
f.location.end will return 369.

Modify the for loop you wrote above (the one which prints the location if the feature type is CDS) - to assign the starting position of the CDS to the variable cds_start and to assign the ending position of the CDS to the variable cds_end.

In [4]:
for f in seq_record.features: # for each feature in this records features
    # ask if the feature is a CDS
        # if so, assign the start location to the variable cds_start
        # if so, assign the end location to the variable cds_end

# print the value of cds_start
# print the value of cds_end


Use the variables cds_start and cds_end to slice the mRNA sequeunce of your gene into three pieces: the 5' untranslated region (fiveUTR_seq), the coding sequence (CDS_seq), and the 3' UTR sequence (threeUTR_seq).

Remember that the nucleotide sequence is found in seq_record.seq.

In [5]:
# assign the nucleotide sequence of your mRNA (seq_record.seq) to a variable mRNA_seq
# print the length of the mRNA sequence

# slice mRNA_seq to return the portion of the mRNA sequence before the begining of the coding sequence, assign this sequence to a variable fiveUTR_seq
# print the length of the 5' UTR sequence

# slice mRNA_seq to return the portion of the mRNA sequence that is the coding sequence, assign this sequence to a variable CDS_seq
# print the length of the coding sequence

# slice mRNA_seq to return the portion of the mRNA sequence after the end of the coding sequence, assign this sequence to a variable threeUTR_seq
# print the length of the 3' UTR sequence

assert len(fiveUTR_seq) + len(CDS_seq) + len(threeUTR_seq) == len(mRNA_seq) #check that the sum of the lengths of the 5'UTR, the coding region, and the 3'UTR are equal to the total length of the mRNA


Now that you've idenfitied the sequences corresponding to different parts of your gene - let's save these sequences as fasta files so that you can access them later.

In [17]:
five_UTR_filename = GI_id + "_5UTR_seq.fasta" # make a name for the file where you'll save your gene's 5'UTR sequence
CDS_filename = GI_id + "_CDS_seq.fasta" # make a name for the file where you'll save your gene's CDS
three_UTR_filename = GI_id + "_3UTR_seq.fasta" # make a name for the file where you'll save your gene's 3'UTR sequence

with open(five_UTR_filename, "w") as fivefile: # open a text file to write to ("w") and call it fivefile
    header = ">" + GI_id + "_5UTR" + '\n' # create a fasta header describing the sequence that will follow
    fivefile.write(header) # write the header to file
    fivefile.write(str(fiveUTR_seq)) #note that the write funciton only accepts strings, so we need to convert the Biopython object containing the sequence to a string


# open a text file to write to ("w") and call it cdsfile
    # create a fasta header describing the sequence that will follow
    # write the header to file
    #note that the write funciton only accepts strings, so we need to convert the Biopython object containing the sequence to a string
    

# open a text file to write to ("w") and call it threefile
    # create a fasta header describing the sequence that will follow
    # write the header to file
    #note that the write funciton only accepts strings, so we need to convert the Biopython object containing the sequence to a strin

That's all folks!  
Please save your notebook and upload the notebook and your .fasta files to Blackboard. Be sure to save ALL of your files as you will be using this notebook for your final project and you'll use the .fasta files generated during this lab to complete next week's lab exercise.